spec/Candid.md

```diff
@@ -711,7 +905,33 @@ not (null <: <datatype>)
 <datatype> <: <datatype'> ~> f
 ------------------------------------------
 <datatype> <: opt <datatype'> ~> \x.?(f x)
+
+not (<datatype> <: <datatype'>)
+---------------------------------
+opt <datatype> <: opt <datatype'>
+  ~> \x.join_opt (\y.
+       if (exists <datatype''>.
+          y : <datatype''> /\
+          <datatype''> <: <datatype> /\
+          <datatype''> <: <datatype> ~> f)
+       then ?(f y)
+       else null)
+```
+where
 ```
+v : <datatype>  iff  M(v : <datatype>) is defined
+```
+The last rule *optimistically* tries to decode an option value when the types are not statically known to match.
+It succceeds if there would have been a valid type for the value (written `v : t`) that is a subtype of both types.
+The effect of this rule is that decoders do not need to perform a subtype check during deserialisation.
+Instead, they can simply try to deserialise, and if they encounter a mismatch abort, to the innermost option type, returning `null`.
+
+*Note:* The working assumption is that the type describing its incoming value is typically *principal*, i.e., the most precise type assignable to the value (in a sense that could be made precise).
```

Comment:
I don’t see right now where we make that assumption, or how it comes into play, but it’s early in the morning… Isn’t the purpose of the whole “if exists” to deal with non-most-precise types?

Anyways, the assumption isn’t even possible (not just costly to enforce): Consider:
```
(vec { opt variant { a } ; opt  variant { a }  }) : (vec ?(variant { a; b }))
```
Clearly the whole message has principal type, but when we reach the optional value `v = opt variant { a }` and want to apply this rule, the type of that value is `?(variant { a; b })` which is not principal, and there is no way to have it principal.

So maybe this Note can be removed? Or maybe I don’t see the point yet.

Reply:
Wording was probably misleading, tweaked. The "principality" property only needs to hold for the whole type, not inductively for its parts. In your example (I suppose the second value ought to be `b`), `?(variant { a; b })` is the principal type -- and it's enough, since it ensures that both `a` and `b` will ultimately be checked against the respective supertype.

Reply:
Hmm, I think I am beginning to understand why you want this principality here.

This might all be clearer if the rules for decoding would just not look at the input type at all, and only at the values.